### DISCUSSION TITLE
Rename "Acceptance" to "Browser Tests" and "Functional" to "Framework Tests"

### DISCUSSION THREAD

**Discussion by ThomasLandauer** [Codeception/Codeception#5970] [Ideas]
While writing https://github.com/Codeception/Codeception/issues/5727#issuecomment-688762499 I started to think about names in general.

* I never understood Codeception's term "Acceptance". In the rest of the world, acceptance testing means if the *user* accepts something, see e.g. https://en.wikipedia.org/wiki/Acceptance_testing But in Codeception it means, if it works in the browser. So I'm suggesting to make this more self-explanatory and rename "Acceptance Tests" to **Browser Tests**

* "Functional Testing", in the rest of the world, is just a catch-all term to check if it works: https://en.wikipedia.org/wiki/Functional_testing In Codeception, the Functional Suite requires a framework, and tests the output of that framework. So I'm suggesting to rename it to **Framework Tests**

Imagine you're explaining the power of Codeception to a friend. And you say: "Besides unit testing, it can also do acceptance and functional tests." What will the friend think that "acceptance" and "functional" refers to??  
However, if you say "...it can also do browser and framework tests", they will at least have a clue what we're talking about...

* BTW: "Unit Testing" isn't a very good term either. "Unit" refers to something like a class: https://en.wikipedia.org/wiki/Unit_testing But I'm *not* suggesting to change it ;-)

**Comment by DavertMik** [comment]
I agree that those terms in a way they transformed in Codecetion are more close to what you propose.
And yes, I also explain these layers as you do - functional=framework, acceptance=browser
however, I am against renaming them for historical and BC reasons.

It was so hard to teach this concept over years and I'd like not to give up on this

P.S. By unit in Codeception we mean both unit & integration tests, so it is even more confusing 😁

**Comment by ThomasLandauer** [comment]
> It was so hard to teach this concept over years

So you prefer to continue the hard way for the upcoming years too?! ;-)

BC break: Yeah, sure I'm talking about v5.0. Suggestion: Make the suite names configurable in `codeception.yml`, with the current names as the default. But starting in v5, set the new names there *for new installations*. So it wouldn't be a hard BC break, but gently moves the users towards the new names.

**Comment by ghost** [comment]
The first time i read the Codeception documentation i had the urge to add comments to my `*.suite.yml` to explain this and not lose sight of this in the distant future.

> functional=framework, acceptance=browser, unit=unit/integration

Now, the Codeception docs has the virtue of explaining the concepts of Testing from scratch while explaining how they are applied in Codeception.

I don't agree with renaming them, but I do agree with documenting this. So...

> What will the friend think that "acceptance" and "functional" refers to??

I imagine my friend, reading another very interesting page of the documentation: **_'Codeception directory structure: What, How, and Where'._**

**Comment by DavertMik** [comment]
Ok, @ThomasLandauer almost convinced me. But we need more voices. So I asked this question on Twitter.
Thomas, If I agree on this, will you help to update all documentation parts? 

@TavoNiievez could you expand on how we can improve in "documenting"?

**Comment by ghost** [comment]
@DavertMik well, what i propose is to create a page dedicated to the Directory Structure, located between 'Getting Started' and 'Acceptance Tests'.

Currently, the introduction explains what functional, unit and acceptance tests are, but it is a bit confusing to guess that it also refers to the fact that in Codeception there are directories with the same names. And that although you can write integration test, in the folder where you will do them by default is in 'Unit'

It is more explicit than updating the introduction to include them there, or having to read the other pages where this is partially explained.

again it's just an idea, I may be missing something.

**Comment by ThomasLandauer** [comment]
> Thomas, If I agree on this, will you help to update all documentation parts?

Yeah, sure!!!  
For some transition period, I'd suggest to use titles like "Browser Tests (formerly: Acceptance Tests)", to help *existing* users to find what they're used to.

@TavoNiievez: Without having looked thoroughly into this, I would guess: This doesn't need an entire page, it can be explained on the first page(s), maybe in an overview table.

**Comment by ghost** [comment]
sure, that would also work

**Comment by DavertMik** [comment]
This is fine to me. @Naktibalda what are your thoughts?

**Comment by roslov** [comment]
Why not to split `unit` to `unit` and `integration` too?

Also,  I agree with `browser` tests instead of `acceptance`.

But I would leave `functional` as is because framework tests do not give better explanation.

In addition, it would be good to have `api` tests by default too.

**Comment by ThomasLandauer** [comment]
Just took a look at the integration tests chapter at https://codeception.com/docs/05-UnitTests#Integration-Tests for the very first time ;-)

@DavertMik Who invented the term? And for what reason?

I see it this way: This is no suite, but rather more features for unit tests. Exactly the same as in browser tests (formerly known as acceptance tests :-) you can enable access to the database too. To me, this in fact is one of the two superpowers of Codeception (the other being the cool syntax). So if there's something called "Unit & Integration Tests", the other suite needs to be called "Browser & Integration Tests" (and "Framework & Integration Tests" too, cause you can enable the REST module and do API testing in there as well).

So I'm suggesting: "Promote" the term "integration tests" from the unit suite up to Codeception in general, like this:
> Codeception is an integration testing framework for PHP, featuring browser testing, framework testing and unit testing.

This could e.g. go into `composer.json` as `decription`: https://github.com/Codeception/Codeception/blob/4.1/composer.json#L3

**Comment by gudwin911** [comment]
have somebody of you ever heard about browser testing and framework testing? This sounds like a bad joke, just try to google these newly invented testing levels, try to find them in https://en.wikipedia.org/wiki/Software_testing or in ISTQB Foundations of software testing book and please explain the results. 

> * acceptance testing means if the _user_ accepts something, see e.g. https://en.wikipedia.org/wiki/Acceptance_testing But in Codeception it means, if it works in the browser

in Codeception acceptance we imitate the real user behavior where the user verifies(accepts) something, why we can't use automation as a part of (or as) an acceptance testing while CI/CD process? In automation this level is called end-to-end test (e2e) https://automationpanda.com/2018/08/01/the-testing-pyramid/

> So if there's something called "Unit & Integration Tests", the other suite needs to be called "Browser & Integration Tests" (and "Framework & Integration Tests" too, cause you can enable the REST module and do API testing in there as well)."

e.g. again, in Codeception acceptance we imitate the real user behavior, real user can't perform call to any of app services as qa while testing could do, but it is not an `Integration testing` it is just a some helper for you to get some data or to make test quicker but it is not a goal of the this level of testing. [Integration testing](https://automationpanda.com/2018/08/01/the-testing-pyramid/) level here is called functional, but for sure it is more understandable than framework testing, and mostly it is used for testing some `api` request/response stuff, but also it can be used like it is written in https://codeception.com/docs/04-FunctionalTests

> Exactly the same as in browser tests (formerly known as acceptance tests :-) you can enable access to the database too.

looks like it is explained [here](https://codeception.com/docs/06-ModulesAndHelpers) and [here](https://codeception.com/docs/07-AdvancedUsage)

@DavertMik please leave current terminology, it is acceptable and understandable for any qa who has heard about testing approaches

**Comment by DavertMik** [comment]
Thanks @gudwin911
Looks like this is a debate from the different sightpoints - from QA and from developer's perspective. From a developer's view everything except unit tests is weird 😁

Because of flexibilty of terms and modules it is to define and hardcode these suites. Maybe E2E is better than acceptance test. Nowadays I find E2E tests much more associated with browser tests.

Regarding to comment of @ThomasLandauer

What I thought of that some Integration tests, which use framework modules, can be thought to be Framework tests! So there is confusion where to put them. We can put all framework tests together into one suite, but those tests will be absolutely different on style (scenario driven VS classical) and on level of pyramid (functional tests rely on Integration tests).

So, there are still more doubts here.

**Comment by ThomasLandauer** [comment]
Let me recap. The general question is:
* Do you want self-explanatory suite names that give **new users** a hint about what they are? The existing terms "acceptance" and "functional" are just arbitrary words that give zero hint, and need to be explained - that's why I created the overview table on top of https://codeception.com/docs/01-Introduction
* Or do you want to keep the existing names **existing users** are used to.

Anybody participating in this discussion is an existing user, since new users certainly won't find their way to an issue number 5970 of a library they're just starting to take a look at...

Answers to other comments:
* "E2E" is just as meaningless as "acceptance". It also doesn't convey the inherent characteristics of acceptance tests - which is (surprise, surprise): browser. In other words: *A functional test is E2E too* - just without JavaScript.
* Unit / Integration tests: The defining characteristics of a unit test is that it tests a single class (="unit") *without context.* This holds true for the Codeception unit suite. Even if it uses some other (framework-)modules to e.g. check the database, that still doesn't make it a "full" test for the entire request-response workflow (like acceptance and functional). In other words: All three Codeception suites can be "integrated" by loading modules.

**Comment by gudwin911** [comment]
> * _A functional test is E2E too_ - just without JavaScript

Agree with you, in current case

> * In other words: All three Codeception suites can be "integrated" by loading modules

Also agree, but isn't it clear that on higher testing level we can use modules from lower levels?

> * "E2E" is just as meaningless as "acceptance". It also doesn't convey the inherent characteristics of acceptance tests - which is (surprise, surprise): browser.

If currently we are still talking about understanding of terminology for the user who has never used codeception and wants to write, in current technology, functional and acceptance test, he can easily find an explanation in documentation or just googling it. But in case he tries to Google explanation of browser or framework test will he find what these testing types mean? (Surprise, surprise) - no. And as far these tests is mostly qa's job, and for them these new naming should be more confusing. But it is just thoughts of one person, do what you want if users support renaming

P.S. also have never seen suggested naming in other testing frameworks (e.g. acceptance, e2e, UI, GUI... but never browser; functional, API, but never framework)

**Comment by eXorus** [comment]
In my project I have 4 suites:
- acceptance (my browser tests)
- api (api tests)
- bdd (my browser tests but only the critical paths written in gherkin)
- integration (ajax call)


I agree that acceptance suite is not understandable first when I present my test framework. I have to say it's the main suite with all the browser tests.

It doesn't matter for me to change acceptance to browser.
